File: DataStructure/BinaryTree.java

/**************************************************
Name: Java Implementation For Binary Tree
Info: http://algorithms.tutorialhorizon.com/breadth-first-searchtraversal-in-a-binary-tree/
Info: http://cslibrary.stanford.edu/110/BinaryTrees.html#java

**************************************************/
import java.util.LinkedList;
import java.util.Queue;
import java.util.Stack;

class Node {
	int data;
	Node left;
	Node right;

	public Node(int data) {
		this.data = data;
		this.left = null;
		this.right = null;
	}
}

class BinaryTree {
	/**************************************************
	Name: Breadth First Search for Binary Tree
	Info: http://algorithms.tutorialhorizon.com/breadth-first-searchtraversal-in-a-binary-tree/
	Note: Using a queue to manage FIFO BFS machanism


	**************************************************/

	public void BFS(Node root) {
		Queue<Node> q = new LinkedList<Node>();
		if (root == null)
			return;
		q.add(root);
		while (!q.isEmpty()) {
			Node n = (Node) q.remove();
			System.out.print(" " + n.data);
			if (n.left != null)
				q.add(n.left);
			if (n.right != null)
				q.add(n.right);
		}
	}

	/**************************************************
	Name: Depth First Search for Binary Tree
	Info: http://algorithms.tutorialhorizon.com/breadth-first-searchtraversal-in-a-binary-tree/
	Note: Using a stack data structure as a buffer to manage BFS mechanism


	**************************************************/
	public void DFS(Node root) {
		Stack<Node> s = new Stack<Node>();
		s.add(root);
		while (s.isEmpty() == false) {
			Node x = s.pop();
			if(x.right!=null) s.add(x.right);
			if(x.left!=null) s.add(x.left);			
			System.out.print(" " + x.data);
		}
	}




	/***********************************************
	3. maxDepth() Solution (Java)
	***********************************************
	 Returns the max root-to-leaf depth of the tree. 
	 Uses a recursive helper that recurs down to find 
	 the max depth. 
	***********************************************/ 
	public int FindMaxDepth(Node root) { 
	    return maxDepth(root); 
	}

	private int maxDepth(Node node) { 
	  if (node==null) { 
	    return(0); 
	  } 
	  else { 
	    int lDepth = maxDepth(node.left); 
	    int rDepth = maxDepth(node.right);

	    // use the larger + 1 
	    return(Math.max(lDepth, rDepth) + 1); //plus this level
	  } 
	} 


	/***********************************************
	4. maxValue() Solution (Java)
	***********************************************
	 Returns the max value in a non-empty binary search tree. 
	 1> Uses a helper method that iterates to the left to find 
	 the min value. 
	 2> Usea recursive method that goes all the way down the right-most node.
	 Since Binary Search Trea is always that larger value is in the right child node.
	***********************************************/ 
	public int findMaxValue(Node root) { 
	 	return maxValue(root); 
	} 
    //iteration version
	
	private int maxValue(Node node) { 
		Node current = node;
		while(current.right != null) {
			current = current.right;
		}

		return(current.data);
	}
	
    /*
	// recursive version
	private int maxValue(Node node) { 
		if(node.right == null)
			return(node.data);
		return (maxValue(node.right));
	}
    */

    /***********************************************
	5. printInOrderTree() Solution (Java)
	***********************************************
	Prints the node values in the "inorder" order. 
    Uses a recursive helper to do the traversal. 
    About what is inorder/postorder: http://www.geeksforgeeks.org/618/
	***********************************************/ 
    public void printInOrderTree(Node root) { 
	    inOrderTree(root); 
	}

	private void inOrderTree(Node root) {
		if (node == null) return;

		 // left, node itself, right 
		 printTree(node.left); 
		 System.out.print(node.data + "  "); 
		 printTree(node.right); 
	}

	/***********************************************
	6. printPostOrderTree() Solution (Java)
	***********************************************
	Prints the node values in the "postorder" order. 
 	Uses a recursive helper to do the traversal. 
	***********************************************/ 
    public void printPostOrderTree(Node root) { 
	    postOrderTree(root); 
	}

	private void postOrderTree(Node node) {
		if (node == null) return;

		 // left, node itself, right 
		 printTree(node.left); 
		 printTree(node.right); 
		 System.out.print(node.data + "  "); 
		 
	}

	/***********************************************
	7. hasPathSum() Solution (Java)
	***********************************************
	Given a tree and a sum, returns true if there is a path from the root 
    down to a leaf, such that adding up all the values along the path 
    equals the given sum.
    Strategy: subtract the node value from the sum when recurring down, 
    and check to see if the sum is 0 when you run out of tree. 
	***********************************************/ 
	public boolean hasPathSum(Node node, int sum) {
		if (node == null) { 
        	return(sum == 0); 
        } 
        else { 
          // otherwise check both subtrees 
            int subSum = sum - node.data; 
            return(hasPathSum(node.left, subSum) || hasPathSum(node.right, subSum)); 
        } 
		

	}




	/***********************************************
	Main Function
	***********************************************/
	public static void main(String[] args) throws java.lang.Exception {
		Node root = new Node(5);
		root.left = new Node(10);
		root.right = new Node(15);
		root.left.left = new Node(20);
		root.left.right = new Node(25);
		root.right.left = new Node(30);
		root.right.right = new Node(35);

		BinaryTree i = new BinaryTree();

		/* BFS */
		//System.out.println("Breadth First Search : ");
		//i.BFS(root);

		/* DFS */
		//System.out.println("Depth-First-Search : ");
		//i.DFS(root);

		/* Min Value */
		//System.out.println("Finding maxValue : ");
		//System.out.println(i.findMaxValue(root));

		/* hasPathSum */
		System.out.println("Tree has path of sum is " + i.hasPathSum(root, 55));


	}
}

File: DataStructure/Trees/BinaryTree.js
"use strict"
const Stack = require('../Lists/Stack');
const Queue = require('../Lists/Queue');
const BinarySearchTree = require("./BinarySearchTree");

function BinaryTree() {
  /**********************************************************************
  Depth First Binary Tree Traversal   using Stack
  **********************************************************************/
  this.depthFirstTranversal = function(root, matches) {
    const nodeStack = new Stack();
    if (root === null)
      return;
    Push(root);

    while (!nodeStack.isEmpty()) {
      const node = nodeStack.pop();
      matches(node.data);
      //console.log(node.data);

      Push(node.left);
      Push(node.right);
    }

    function Push(node) {
      if (node === null || node === undefined)
        return;

      nodeStack.push(node);

    }

    return;
  }



  /**********************************************************************
  Breadth First Binary Tree Traversal  using Queue
  **********************************************************************/
  this.breadthFirstTranversal = function(root, matches) {
    const nodeQueue = new Queue();
    if (root === null)
      return;
    Push(root);

    while (!nodeQueue.isEmpty()) {
      const node = nodeQueue.pop();
      matches(node.data);
      Push(node.left);
      Push(node.right);
    }

    function Push(node) {
      if (node === null || node === undefined)
        return;

      nodeQueue.push(node);
    }

    return;
  }


  /**********************************************************************
  InOrder Binary Tree Traversal  
  **********************************************************************/
  this.inOrderTranversal = function(root, matches) {
    if (root !== null && root !== undefined) {
      //matches(root.data);  preOrder
      this.inOrderTranversal(root.left, matches);
      matches(root.data);
      this.inOrderTranversal(root.right, matches);
      //matches(root.data);  postOrder
    }
  }
}

module.exports = BinaryTree;



//for testing
/*
let bst = new BinarySearchTree();

bst.push(0);
bst.push(5);
bst.push(3);
bst.push(9);
bst.push(6);
bst.push(2);
bst.push(4);
bst.push(1);
bst.push(7); 

//depthFirstSearch(bst.root, (val)=>console.log(val));
let binaryTree = new BinaryTree();

//for testing
const testArray = [9, 8, 7, 6, 5, 4, 3, 2, 1, 0];
bst.createBST(testArray);
binaryTree.breadthFirstTranversal(bst.root, (val) => console.log(val));
console.log("---------------------------------------------");
binaryTree.depthFirstTranversal(bst.root, (val) => console.log(val));
console.log("---------------------------------------------");
binaryTree.inOrderTranversal(bst.root, (val) => console.log(val));
*/
